qutebrowser/browser/webkit/network/pac.py

```diff
@@ -0,0 +1,299 @@
+    exceptions returning them as JavaScript Error objects.
+
+    Args:
+        args: Types of method arguments.
+
+    Return: Wrapped method.
+    """
+    def _decorator(method):
+        @functools.wraps(method)
+        def new_method(self, *args, **kwargs):
+            try:
+                return method(self, *args, **kwargs)
+            except:
+                e = str(sys.exc_info()[0])
+                log.network.exception("PAC evaluation error")
+                # pylint: disable=protected-access
+                return self._error_con.callAsConstructor([e])
+                # pylint: enable=protected-access
+        return pyqtSlot(*args, result=QJSValue)(new_method)
+    return _decorator
+
+
+class _PACContext(QObject):
+
+    """Implementation of PAC API functions that require native calls.
+
+    See http://web.archive.org/web/20060424005037/wp.netscape.com/eng/mozilla/2.0/relnotes/demo/proxy-live.html
+    """
+
+    JS_DEFINITIONS = """
+        function dnsResolve(host) {
+            return PAC.dnsResolve(host);
+        }
+
+        function myIpAddress() {
+            return PAC.myIpAddress();
+        }
+    """
+
+    def __init__(self, engine):
+        """Create a new PAC API implementation instance.
+
+        Args:
+            engine: QJSEngine which is used for running PAC.
+        """
+        super().__init__(parent=engine)
+        self._engine = engine
+        self._error_con = engine.globalObject().property("Error")
+
+    @_js_slot(str)
+    def dnsResolve(self, host):
+        """Resolve a DNS hostname.
+
+        Resolves the given DNS hostname into an IP address, and returns it
+        in the dot-separated format as a string.
+
+        Args:
+            host: hostname to resolve.
+        """
+        ips = QHostInfo.fromName(host)
+        if ips.error() != QHostInfo.NoError or not ips.addresses():
+            err_f = "Failed to resolve host during PAC evaluation: {}"
+            log.network.info(err_f.format(host))
+            return QJSValue(QJSValue.NullValue)
+        else:
+            return ips.addresses()[0].toString()
+
+    @_js_slot()
+    def myIpAddress(self):
+        """Get host IP address.
+
+        Return the server IP address of the current machine, as a string in
+        the dot-separated integer format.
+        """
+        return QHostAddress(QHostAddress.LocalHost).toString()
+
+
+class PACResolver(object):
+
+    """Evaluate PAC script files and resolve proxies."""
+
+    @staticmethod
+    def _parse_proxy_host(host_str):
+        host, _colon, port_str = host_str.partition(':')
+        try:
+            port = int(port_str)
+        except ValueError:
+            raise ParseProxyError("Invalid port number")
+        return (host, port)
+
+    @staticmethod
+    def _parse_proxy_entry(proxy_str):
+        """Parse one proxy string entry, as described in PAC specification."""
+        config = [c for c in proxy_str.split(' ') if c]
```

Comment:
Should't you do `c.strip() for c in ...` here? MDN shows `PROXY w3proxy.netscape.com:8080; PROXY mozilla.netscape.com:8081` as example value, and the caller of `_parse_proxy_entry` only splits on `;` - or am I missing something?